lib/que/worker.rb

```diff
@@ -5,21 +5,67 @@
 require "que/locker"
 
 module Que
+  class JobTimeoutError < StandardError; end
+
   class Worker
     # Defines the time a worker will wait before checking Postgres for its next job
     DEFAULT_QUEUE = ''
     DEFAULT_WAKE_INTERVAL = 5
     DEFAULT_LOCK_CURSOR_EXPIRY = 0 # seconds
+    DEFAULT_STOP_TIMEOUT = 5 # seconds
+
+    # Start worker_count number of threaded workers, returning a proc that can be called
+    # to stop the workers from working.
+    def self.start_workers(worker_count, metrics_labels: {}, **kwargs)
+      Que.logger.info(msg: 'Starting workers', event: 'que.worker.start', worker_count: worker_count)
+
+      # Create each worker with a distinct ID in order to differentiate each workers
+      # prometheus metrics.
+      workers = Array.new(worker_count) do |id|
+        Worker.new(**kwargs.merge(metrics_labels: metrics_labels.merge(worker: id)))
+      end
+
+      worker_threads = workers.map { |worker| Thread.new { worker.work_loop } }
+
+      -> (stop_timeout = DEFAULT_STOP_TIMEOUT) do
```

Comment:
For me, it felt sensible that this method returned a callback that could be used to tidy-up and release resources. I've spent far too much time in Go code recently though, and I'm very aware this may seem surprising to the unsuspecting Ruby-ist.

An alternative would be to create a new construct, `Que::WorkerGroup`, and move the code into that class. The advantage is that you'd have a first class construct to formalize the behaviour of a 'worker group'; disadvantages are an entirely new construct for minimal code, potentially clunky API around `new` and `start` behaviour, and the code that constructs `Worker` instances is placed further away from the definition of the class.

I'm in favour of leaving this here, but would appreciate feedback from the reviewer on whether that makes sense.

Reply:
I don't think there's anything wrong with this per-se. Returning a closure which encapsulates the state of the worker pool is perfectly reasonable in a functional language.

In Ruby, I think the more idiomatic way would be to store the thread collection as instance variables (e.g. on a pool/group) and then ask the pool to stop itself (which would iterate through that internal collection of threads). Having said that, both options "work" and frankly we could implement your group idea at a later date if we felt this was slowing us down. Both options are comprehensible IMO.